### PR TITLE
feat(backend)(rules): add RunValidationService for submitted board sets

### LINKED ISSUE #278: feat(backend)(rules): add run validation service
## Description

Describe the feature briefly.

## Acceptance Criteria

- [ ]

## Notes

Additional context, screenshots, or links.

### LINKED ISSUE #279: feat(backend)(rules): validate run minimum size
## Description

Describe the feature briefly.

## Acceptance Criteria

- [ ]

## Notes

Additional context, screenshots, or links.

### LINKED ISSUE #280: feat(backend)(rules): validate equal color in run
## Description

Describe the feature briefly.

## Acceptance Criteria

- [ ]

## Notes

Additional context, screenshots, or links.

### LINKED ISSUE #281: feat(backend)(rules): validate duplicate numbers in run
## Description

Describe the feature briefly.

## Acceptance Criteria

- [ ]

## Notes

Additional context, screenshots, or links.

### LINKED ISSUE #282: feat(backend)(rules): validate ascending order in run
## Description

Describe the feature briefly.

## Acceptance Criteria

- [ ]

## Notes

Additional context, screenshots, or links.

### LINKED ISSUE #283: feat(backend)(rules): add gap detection in run validation
## Description

Describe the feature briefly.

## Acceptance Criteria

- [ ]

## Notes

Additional context, screenshots, or links.

### PR DESCRIPTION
**Pull request by miri2468**
## Summary

Add the first concrete rule validator in the backend rule-validation flow: a dedicated `RunValidationService` that validates whether one submitted `BoardSet` is a legal **run**. This PR should stay very focused: only **runs**, no groups, no full board validation, and no end-turn integration yet.

## Why this is the best next step

So this PR should establish one thing only:

> "Is this submitted `BoardSet` a valid run?"

## Related Issues

- Related to #277 
- Closes #278 
- Closes #279 
- Closes #280 
- Closes #281 
- Closes #282 
- Closes #283 
- Related to #284 
- Related to #285 
- Related to #286  
- Related to #293
- Related to #289 
- 

## Scope

Included:

- `RunValidationService`
- run-validation checks for non-joker sets
- clear `ValidationResult` / violation handling using the existing rule-validation foundation
- focused unit tests for valid and invalid run cases

Not included:

- group validation
- set validation that tries both group and run
- board validation
- joker-aware run validation
- first-move validation
- end-turn integration
- websocket changes

## What a valid run means in this PR

For this first version, keep the rules strict and simple.

A submitted `BoardSet` counts as a valid **run** only if:

1. it contains at least **3 tiles**
2. all tiles are **numbered tiles** (no jokers in this PR)
3. all tiles have the **same color**
4. all tile numbers are **unique**
5. the numbers form one **strictly ascending consecutive sequence** with no gaps

### Valid examples

- `Red 3, Red 4, Red 5`
- `Blue 10, Blue 11, Blue 12, Blue 13`

### Invalid examples

- `Red 7, Red 8` → too small
- `Red 3, Blue 4, Red 5` → mixed colors
- `Red 3, Red 5, Red 6` → gap in sequence
- `Red 7, Red 7, Red 8` → duplicate number
- `Red 5, Red 4, Red 3` → wrong order if submitted order is expected to already be normalized
- any set containing a joker → reject for now

### 1. Add `RunValidationService`

Create a dedicated backend service that validates exactly one `BoardSet` as a run.

A good method shape is:

```kotlin
fun validate(set: BoardSet): ValidationResult
```

The method should check each run rule and return a result using the existing rule-validation model.

### 2. Keep jokers out of scope

For this PR:

- if the set contains one or more jokers, return invalid
- do not try to substitute joker values yet
- leave joker-aware run validation to a later PR

### 3. Decide how to treat order

For this PR, the run validator should **not trust** the stored tile order inside the `BoardSet`.

Instead:

- the validator should extract the numbered tiles
- sort them by tile number first
- then validate the run rules on that sorted sequence

This means a run should still be accepted even if the tiles were persisted in a mixed order, as long as sorting reveals one valid consecutive same-color sequence.

### 4. Suggested implementation shape

```kotlin
class RunValidationService {
    fun validate(set: BoardSet): ValidationResult {
        if (set.tiles.size < 3) {
            return ValidationResult.invalid("Run must contain at least 3 tiles")
        }

        if (set.tiles.any { it is JokerTile }) {
            return ValidationResult.invalid("Joker support is not implemented yet")
        }

        val numberedTiles = set.tiles.map { it as NumberedTile }

        val distinctColors = numberedTiles.map { it.color }.distinct()
        if (distinctColors.size != 1) {
            return ValidationResult.invalid("All run tiles must have the same color")
        }

        val sortedTiles = numberedTiles.sortedBy { it.number }
        val numbers = sortedTiles.map { it.number }
        if (numbers.size != numbers.distinct().size) {
            return ValidationResult.invalid("Run tiles must not contain duplicate numbers")
        }

        for (i in 1 until numbers.size) {
            if (numbers[i] != numbers[i - 1] + 1) {
                return ValidationResult.invalid("Run tiles must form a consecutive ascending sequence")
            }
        }

        return ValidationResult.valid()
    }
}
```

If your current rule-validation foundation already uses `RuleViolation` objects, prefer structured violations over plain string-only invalid results.

## Checks this validator should perform

In a readable order:

1. minimum size
2. joker not supported yet
3. all numbered tiles share the same color
4. all numbers are unique
5. after sorting by number, the tiles form a strictly ascending consecutive sequence

## Tests to include

Add focused unit tests for `RunValidationService`.

At minimum:

- valid 3-tile run with same color and consecutive ascending numbers
- valid longer run with same color and consecutive ascending numbers
- rejects run with fewer than 3 tiles
- rejects run with mixed colors
- rejects run with duplicate numbers
- rejects run with a gap in the number sequence
- rejects run containing jokers
- accepts a valid run even if the tiles are submitted in a mixed order
- rejects a mixed-order set if sorting still does not produce a valid consecutive run

## Expected outcome of this PR

After this PR, the backend should be able to validate one submitted `BoardSet` as a possible **run** and clearly answer:

- valid run
- invalid run, with a clear failure reason

That gives the next PR a clean base for:

- later `SetValidationService`
- later `BoardValidationService`

### CODE REVIEW

File: apps/backend/src/test/kotlin/at/se2group/backend/rules/service/RunValidationServiceTest.kt [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    fun `reject run if tiles are less than 3`() {
+        val result = service.validate(
+            set(
+                tile("tile-1", TileColor.RED, 7),
+                tile("tile-2", TileColor.RED, 8)
+            )
+        )
+        assertFalse(result.isValid)
+        assertEquals("RUN_MIN_SIZE", result.violations.single().code)
+        assertEquals("Run must contain at least 3 tiles", result.violations.single().message)
```
Comment (kingjulien1) [inline apps/backend/src/test/kotlin/at/se2group/backend/rules/service/RunValidationServiceTest.kt]: One thing still missing in the tests: they do not verify the returned `tileIds` and `boardSetId`, even though both are part of the actual validation result contract here.

Since those fields are important for frontend-side highlighting and mapping the violation back to the correct board set, I’d add assertions for them as well — otherwise a future regression could silently break that context without failing these tests.

For example:

```kotlin
val violation = result.violations.single()
assertEquals(set.boardSetId, violation.boardSetId)
assertEquals(set.tiles.map { it.tileId }, violation.tileIds)
```

File: apps/backend/src/test/kotlin/at/se2group/backend/rules/service/RunValidationServiceTest.kt [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        val result = service.validate(
+            set(
+                tile("tile-1", TileColor.RED, 5),
+                tile("tile-2", TileColor.RED, 4),
+                tile("tile-3", TileColor.RED, 3)
+            )
+        )
+        assertTrue(result.isValid)
+        assertTrue(result.violations.isEmpty())
+    }
```
Comment (kingjulien1) [inline apps/backend/src/test/kotlin/at/se2group/backend/rules/service/RunValidationServiceTest.kt:139]: This is a good test, but I think it is still a bit weaker than it could be.

`5, 4, 3` proves that descending order is accepted after sorting, but I’d also add one genuinely shuffled valid case like `4, 6, 5`. That would pin down more clearly that the validator really treats persisted order as irrelevant and validates based on the sorted sequence.

File: apps/backend/src/test/kotlin/at/se2group/backend/rules/service/RunValidationServiceTest.kt [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                tile("tile-1", TileColor.RED, 7),
+                tile("tile-2", TileColor.RED, 7),
+                tile("tile-3", TileColor.RED, 8)
+            )
+        )
+        assertFalse(result.isValid)
+        assertEquals("RUN_DUPLICATE_NUMBER", result.violations.single().code)
+        assertEquals("Run tiles must not have duplicate numbers", result.violations.single().message)
+    }
+
```
Comment (kingjulien1) [inline apps/backend/src/test/kotlin/at/se2group/backend/rules/service/RunValidationServiceTest.kt:109]: I think one useful missing test here is validation precedence between duplicate-number failure and non-consecutive failure.

Since the service checks duplicate numbers before sequence continuity, a case like `3, 3, 5` should clearly return `RUN_DUPLICATE_NUMBER`, not `RUN_NOT_CONSECUTIVE`. I’d add that explicitly so the validator order is locked down by tests.

**Review by kingjulien1 — CHANGES_REQUESTED** [review]
Great job overall 👍 — the service and tests already form a solid first rules PR, and I only left a few smaller comments to tighten the edges a bit.

File: apps/backend/src/main/kotlin/at/se2group/backend/rules/service/RunValidationService.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        for (i in 1 until sortedNumbers.size) {
+            if (sortedNumbers[i] != sortedNumbers[i - 1] + 1) {
+                return invalid(
+                    code = "RUN_NOT_CONSECUTIVE",
+                    message = "Run tiles must create a consecutive ascending sequence",
+                    tileIds = numberedTiles.map { it.tileId }
+                )
+            }
```
Comment (kingjulien1) [inline apps/backend/src/main/kotlin/at/se2group/backend/rules/service/RunValidationService.kt]: a bit cleaner version

```kotlin
val isConsecutive = sortedNumbers
        .zipWithNext()
        .all { (a, b) -> b == a + 1 }

    if (!isConsecutive) {
        return invalid(
            code = "RUN_NOT_CONSECUTIVE",
            message = "Run tiles must create a consecutive ascending sequence",
            tileIds = numberedTileIds
        )
    }
```

File: apps/backend/src/main/kotlin/at/se2group/backend/rules/service/RunValidationService.kt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        val numberedTiles = set.tiles.map { it as NumberedTile }
+
+        val distinctColors = numberedTiles.map { it.color }.distinct()
+            if (distinctColors.size != 1 ) {
+                return invalid(
+                    code = "RUN_COLOR_MISMATCH",
+                    message = "All run tiles must have the same color",
+                    tileIds = numberedTiles.map { it.tileId }
+                )
+            }
```
Comment (kingjulien1) [inline apps/backend/src/main/kotlin/at/se2group/backend/rules/service/RunValidationService.kt:43]: same thing again here: 
small formatting issue: please shift the `if`-block to the left to align with the base method alignment.

File: apps/backend/src/main/kotlin/at/se2group/backend/rules/service/RunValidationService.kt [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+class RunValidationService {
+
+    fun validate(set: BoardSet): ValidationResult {
+            if (set.tiles.size < 3) {
+                return invalid(
+                    code = "RUN_MIN_SIZE",
+                    message = "Run must contain at least 3 tiles",
+                    tileIds = set.tiles.map { it.tileId }
+                )
+            }
```
Comment (kingjulien1) [inline apps/backend/src/main/kotlin/at/se2group/backend/rules/service/RunValidationService.kt:22]: small formatting issue: please shift the if-block to the left to align with the base method alignment.

File: apps/backend/src/main/kotlin/at/se2group/backend/rules/service/RunValidationService.kt [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+@Service
+class RunValidationService {
+
+    fun validate(set: BoardSet): ValidationResult {
+            if (set.tiles.size < 3) {
+                return invalid(
+                    code = "RUN_MIN_SIZE",
+                    message = "Run must contain at least 3 tiles",
+                    tileIds = set.tiles.map { it.tileId }
+                )
```
Comment (kingjulien1) [inline apps/backend/src/main/kotlin/at/se2group/backend/rules/service/RunValidationService.kt:21]: please include the setIndex in every `invalid` call from here on out.

**Review by kingjulien1 — CHANGES_REQUESTED** [review]
just some small little follow-ups that I forgot to include in the first review from earlier.

**Review by kingjulien1 — APPROVED** [review]
Really nice work! 👍 Thanks for implementing all my tedious, messy review comments and thanks for resolving the GitHub comment conversations - I really appreciate it a lot. 

You can merge now 👍 don't forget to delete the source branch
